# Machine Learning and Statistics - Tasks
Assignment Tasks for Machine Learning and Statistics, GMIT 2020

Lecturer: dr Ian McLoughlin


>Author: **Andrzej Kocielski**  
>Github: [andkoc001](https://github.com/andkoc001/)  
>Email: G00376291@gmit.ie, and.koc001@gmail.com

Created: 05-10-2020

This Notebook should be read in conjunction with the corresponding `README.md` file at the project [repository](https://github.com/andkoc001/Machine-Learning-and-Statistics.git) at GitHub.

___
## Task 1

Objectives: __Print on screen the square root of 2 to 100 decimal places__.

### Division of a range method

For any real number $ x $, that $ x > 1 $:

$$ \sqrt{x} \cdot \sqrt{x} = x $$

$$ 1 < \sqrt{x} < x $$

The last equation is an equivalent to $ 1^2 < x < x^2 $.


Hence, it is possible to approximate the value of $ \sqrt{x} $ by iteratively testing into which of the halves of the original range will it fall. This is done by performing the test: 

$$ (\sqrt{x})^2 < (\frac{1+x}{2})^2 $$

Then in the next iteration new boundary conditions are assumed. If the test is true, $ \frac{1+x}{2} $ becomes the right boundary; if the test is false, $ \frac{1+x}{2} $ becomes the left boundary. This way, the range tightens at each iteration. 


___
Example for $ x = 2 $:

The initial conditions is this: $ 1^2 < (\sqrt{2})^2 < 2^2 $.

In the first iteration, the left boundary is $ 1^2 = 1 $, and the right boundary is $ 2^2 = 4 $. 

Then we perform the test: $ (\frac{1+2}{2})^2 = 2.25 $, which is greater than $ (\sqrt{2})^2 = 2 $.  

Therefore, in the second iteration the left boundary remains $ 1^2 = 1 $, and the right boundary becomes $ \frac{1+2}{2} = 1.5 $.

We do the test again: $ (\frac{1+1.5}{2})^2 = 1.5625 $. This is less than $ (\sqrt{2})^2 = 2 $. 

In the third iteration the left boundary becomes $ \frac{1+1.5}{2} = 1.25 $, and the right boundary stays $ \frac{1+2}{2} = 1.5 $.

We do the test again: $ (\frac{1.25+1.5}{2})^2 = 1.890625 $. This is less than $ (\sqrt{2})^2 = 2 $. 

In the forth iteration the left boundary becomes $ \frac{1.25+1.5}{2} = 1.375 $, and the right boundary stays $ \frac{1+2}{2} = 1.5 $.

And so on...
___

This process may continue until required precision is achieved. 

For Python built-in data types, _while_ loop may govern the precision improvement process. However, Python allows only precision only to 15 digits precision.

Let's designate the required precision as $ \tau $. As long as $ (\frac{1+x}{2})^2 >= \tau $, the required precision is not achieved and another iteration is to be performed.

In [1]:
# Define number of which sqare root will be approximated
number = 2

# Define decimal places precision
precision = 15 # fiveteen decimal places appears to be the maximum for this data type

# Initial boundary conditions:
left = 1
right = number
middle = (left+right) / 2

# Implementing the logic
iteration = 0 

# Loop exit condition, i.e. requested precision is achieved
while abs(number-middle*middle) >= 10**-precision:
    
    # Testing which half of the range the square root of the number will fall into; tightening the boundaries
    if middle*middle > number:
        # if the condition is satisfied, the right boundary is changed to the previous mid-point
        right = middle
        
    else:
        # if the condition is not satisfied, the left boundaries is changed to the previous mid-point
        left = middle
    
    # Update the value of the variable 'middle'
    middle = (left+right) / 2
    
    # Update number of iteration
    iteration = iteration + 1

    # Print out intermediate results for each iteration
    # print(f"Iteration {iteration:3}, left: {left:18}, Sqare root: {middle:.16}, \tright: {right:18}")


# Print out the result
print(f"Iteration {iteration:3}, The sqare root of {number}: {middle:.16}")  


Iteration  49, The sqare root of 2: 1.414213562373095


### Binary shift with Newton's method
Adapted from https://stackoverflow.com/a/28151578

> `>> 1` is a bit-wise right shift, e.g. "divide by 2", `>> 2` would be "divide by 4", `>> n` is "divide by 2**(n)" - https://stackoverflow.com/users/118068/marc-b

In [2]:
### Method C - Newton's method
# Adapted from https://stackoverflow.com/a/28151578# Adapted from https://stackoverflow.com/a/28151578
''' Long integer square roots. Newton's method.
    Written by PM 2Ring. Adapted from C to Python 2008.10.19
'''

# the algorithm

def root(m):
    
    # Get initial approximation
    n, a, k = m, 1, 0
    
    while n > a:
        n >>= 1
        a <<= 1
        k += 1
        #print('\', k, ':', n, a) # for debugging

    # Go back one step & average
    a = n + (a>>2)
    
    #print(a) # for debugging

    # Apply Newton's method
    
    while k:
        a = (a + m // a) >> 1
        k >>= 1
        #print(k, ':', a) # for debugging
    
    result = a
    return result

# body of the main function
def main():
    
    # number to be square rooted, between 1 and 99 - outside the range there is possible error notation, beyond the scope of this task
    number = 2
    
    # number of decimal places to be shown
    precision = 100
    
    factor  = 10 ** (number * precision)
    m =  number * factor
    
    # print the result converted to a string
    string_result = str(root(m))
    
    # Check if the number is in the requested range
    if 1 <= number < 100:
        
        # take the first digit followed by a dot
        result = string_result[0:1] + "."
        
        # take the remaining digits up to set precision
        for i in string_result[1:precision]:
            result = result + i
    
        print("The Square Root of " + str(m/factor) + ":")
        print(result)
                
    else:
        print("Choose number to be squared between 1 and 99")
        
# execute only if run as a script
if __name__ == '__main__':
    main()

The Square Root of 2.0:
1.414213562373095048801688724209698078569671875376948073176679737990732478462107038850387534327641572


### Square root by subtraction method

In [3]:
# The algorithm "Square root by subtraction" by Frazer Jarvis, can be found here: http://www.afjarvis.staff.shef.ac.uk/maths/jarvisspec02.pdf

# Adapted from https://www.mathblog.dk/project-euler-80-digits-irrational-square-roots/

def Squareroot(n, prec):

    # Set the required parameters
    limit = 10**(prec+1) # extra precision to avoid rounding error
    left = 5 * n
    right = 5
 
    while right < limit:
        
        if left >= right:
            left -= right
            right += 10
            
        else:
            left *= 100
            right = 10 * right - 45;
       
    return right;



# body of the main function
def main():
    
    # number to be square-rooted, between 1 and 99 - outside the range there is possible error notation, beyond the scope of this task
    number = 2
    
    # number of decimal places to be shown
    precision = 100
    
    # call the main algorithm and print the result converted to a string
    string_result = str(Squareroot(number, precision))
    
    # Check if the number is in the requested range
    if 1 <= number < 100:
        
        # take the first digit followed by a dot
        result = string_result[0:1] + "."
        
        iteration = 0      
        # take the remaining digits up to set precision
        for i in string_result[1:precision]:
            result = result + i
            iteration += 1
            # print(f"Iter: {iteration:3}: {result}") # for debugging
               
        print(len(result))
        print("The Square Root of " + str(number) + ":")
        print(result)
                
    else:
        print("Choose number to be squared between 1 and 99")
        
        
        
# execute only if run as a script
if __name__ == '__main__':
    main()

101
The Square Root of 2:
1.414213562373095048801688724209698078569671875376948073176679737990732478462107038850387534327641572


#### Result verification
The above result can be checked against the approximation of $\sqrt{2}$ available, amoong others, at https://apod.nasa.gov/htmltest/gifcity/sqrt2.1mil. Here, the first 102 digits (including that on the left of the decimal) is shown:

$ \sqrt{2} $= 1.4142 13562  37309 50488  01688 72420  96980 78569  67187 53769  48073 17667  97379 90732  47846 21070  38850 38753  43276 41572 73 

___
## References and bibliography 

### General 

- Ian McLoughlin, Assignment Brief, 2020. [pdf] GMIT. Available at: <https://learnonline.gmit.ie/mod/url/view.php?id=102004> [Accessed October 2020].
- Ian McLoughlin, Lecturer's notes on square root of 2, 2020. [pdf] GMIT. Available at: <https://learnonline.gmit.ie/mod/url/view.php?id=92022> [Accessed October 2020].

### Task 1 related

- Ian McaLoughlin - Introduction to the task #1, 2020 [online]. Available at <https://github.com/ianmcloughlin/playing-with-jupyter/blob/main/playing-with-jupyter.ipynb> [Accessed October 2020]
- Taylor series - Wikipedia. [online] Available at: <https://en.wikipedia.org/wiki/Taylor_series> [Accessed October 2020].
- Mateusz Kowalski - Matematyka, Wzór Taylora i Maclaurina, Przybliżanie Funkcji (in Polish) [online]. Available at: http://www.kowalskimateusz.pl/matematyka-wzor-taylora-i-maclaurina-przyblizanie-funkcji/ [Accessed October 2020]
- The Penn Calc Wiki, Taylor Series [online]. Available at: <http://calculus.seas.upenn.edu/?n=Main.TaylorSeries> [Accessed October 2020]
- NASA - Square root of 2 - the first million digits [online]. Available at: <https://apod.nasa.gov/htmltest/gifcity/sqrt2.1mil> [Accessed October 2020]
- Python manual on Decimal library [online]. Available at: <https://docs.python.org/3/library/decimal.html> [Accessed October 2020]
- Python manual on BitwiseOperations [online]. Available at: <https://wiki.python.org/moin/BitwiseOperators> [Accessed October 2020]
- Stack Overflow - find as many digits of the square root of 2 as possible [online]. Available at: <https://stackoverflow.com/a/15434306> [Accessed October 2020]
- Frazer Jarvis, 2005 - Square root by subtraction [pdf]. Available at: <http://www.afjarvis.staff.shef.ac.uk/maths/jarvisspec02.pdf> [Accessed October 2020]

___
Andrzej Kocielski